Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -2271,6 +2371,10 @@ protected class ClassWriter : IClassWriter {
       to = to.NormalizeExpand();
       Contract.Assert(from.IsRefType == to.IsRefType);
 
+      if (from.IsIndDatatype || from.IsCoDatatype) {
```

Comment:
`from.IsIndDatatype || from.IsCoDatatype` can be written as just `from.IsDatatype`.